packages/server/parseFormBody.ts

```diff
@@ -0,0 +1,52 @@
+import {APP_MAX_AVATAR_FILE_SIZE} from 'parabol-client/utils/constants'
+import {HttpResponse, getParts, MultipartField} from 'uWebSockets.js'
+
+interface Uploadable {
+  contentType: string
+  buffer: Buffer
+}
+
+type ParseFormBodySignature = {
+  res: HttpResponse
+  contentType: string
+}
+
+const parseRes = (res: HttpResponse) => {
+  return new Promise<Buffer | null>((resolve) => {
+    let buffer: Buffer
+    res.onData((ab, isLast) => {
+      const curBuf = Buffer.from(ab)
+      buffer = buffer ? Buffer.concat([buffer, curBuf]) : isLast ? curBuf : Buffer.concat([curBuf])
+      if (buffer.length > APP_MAX_AVATAR_FILE_SIZE) resolve(null)
+      if (isLast) resolve(buffer)
+    })
+  })
+}
+
+const parseFormBody = async ({res, contentType}: ParseFormBodySignature): Promise<JSON | null> => {
+  const resBuffer = await parseRes(res)
+  if (!resBuffer) return null
+  const parts = getParts(resBuffer, contentType) as MultipartField[]
+  if (!parts) return null
+
+  try {
+    let parsedBody
+    parts.forEach(({name, data, type}) => {
+      if (name === 'body') {
+        parsedBody = JSON.parse(Buffer.from(data).toString())
+        return
+      }
+      if (!name.startsWith('uploadables')) return
+      const [, key] = name.split('.')
+      parsedBody.payload.variables[key] = {
```

Comment:
we can't be sure that `parsedBody.payload` exists, or that it includes variables or that it comes before the `uploadable` key does.
we want to get all the keys & put them in an object & then after the loop, put that object inside the parsedBody iif it exists.

Reply:
@mattkrick is it wrong to assume that all http requests going to this handler (httpGraphQLHandler) will be the type below? i'm assuming all requests should conform to this interface, which would mean that `body.payload` should exist. if not, the server will return 422 client error.

```
interface FetchHTTPData {
  type: 'start' | 'stop'
  payload: OperationPayload
}
```

i am also targeting `variables` instead of `parsedBody` so that the uploadable will be made available to the mutation resolver as a variable. the other option was putting it in context, but i thought this way was cleaner. There is the possibility a variable might accidentally get overwritten, but only if the developer passed an uploadable under the same name key as another passed variable. i think the mistake would probably throw an error bc of type checking, which seems okay imo. what do you think?